actioncable/app/assets/javascripts/action_cable/connection.coffee

```diff
@@ -89,19 +97,49 @@ class ActionCable.Connection
 
     open: ->
       ActionCable.log("WebSocket onopen event")
-      @disconnected = false
-      @subscriptions.reload()
+      # This is in the case that the browser client opens
+      # the connection regardless of whether a protocol was 
+      # negotiated successfully or not.
+      if @isProtocolSupported()
+        ActionCable.log("WebSocket using '#{@getProtocol()}' subprotocol.")
+        @disconnected = false
+        @subscriptions.reload()
+      else
+        @close()
 
-    close: ->
+    close: (event) ->
       ActionCable.log("WebSocket onclose event")
-      @disconnect()
+
+      # According to https://www.w3.org/TR/websockets/#concept-websocket-close-fail
+      # a sub-protocol mismatch should be treated as a 1006 error, indistinguishable
+      # from any other failure to connect. Not all browsers behave this way:
+
+      # This encompases the behavior of most browsers when it comes to
+      # reacting to a failed protocol negotiation.
+      if event.code is 1000 and event.wasClean and @getProtocol() is ""
+        @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
+
+      # This is Chrome's response (which is technically to spec) and looks 
+      # exactly the same as a normal bad disconnect. Thus, we must test to 
+      # see if the server would connect under normal conditions. If so, Chrome reacted
+      # to the failed protocol negotiation, and not something else.
+      else if event.code is 1006 and not event.wasClean and @getProtocol() is ""
+        testSocket = new WebSocket(@consumer.url)
+        testSocket.onopen = =>
+          @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
+          testSocket.close()
```

Comment:
For 1. This is part of the WebSocket spec. If the server doesn't support any protocols, it needs to not return that header or return the header with a null value. In our case, there is no visibility into this process [as the WebSocket driver does this for us](https://github.com/faye/websocket-driver-ruby/blob/master/lib/websocket/driver/hybi.rb#L76).

Re 4: This is a fantastic idea! I'm going to give this a try. I think this would do the trick.